jetstream/analysis.py

```diff
@@ -242,30 +246,85 @@ def _calculate_statistics(self, metrics_table: str, period: AnalysisPeriod):
-            "Analysis.run invoked for experiment %s", self.config.experiment.normandy_slug
-        )
-
+    def is_runnable(self, current_date: Optional[datetime] = None) -> bool:
         if self.config.experiment.normandy_slug is None:
             self.logger.info("Skipping %s; no slug", self.config.experiment.normandy_slug)
-            return  # some experiments do not have a normandy slug
+            return False  # some experiments do not have a normandy slug
 
         if not self.config.experiment.proposed_enrollment:
             self.logger.info(
                 "Skipping %s; no enrollment period", self.config.experiment.normandy_slug
             )
-            return
+            return False
 
         if self.config.experiment.start_date is None:
             self.logger.info("Skipping %s; no start_date", self.config.experiment.normandy_slug)
-            return
+            return False
 
-        if self.config.experiment.end_date and self.config.experiment.end_date < current_date:
+        if (
+            current_date
+            and self.config.experiment.end_date
+            and self.config.experiment.end_date < current_date
+        ):
             self.logger.info("Skipping %s; already ended", self.config.experiment.slug)
+            return False
+
+        return True
+
+    def validate(self) -> None:
+        if not self.is_runnable():
+            raise Exception("Cannot validate experiment")
+
+        dates_enrollment = self.config.experiment.proposed_enrollment + 1
+
+        analysis_length_dates = (
+            (self.config.experiment.end_date - self.config.experiment.start_date).days
+            - dates_enrollment
+            + 1
+        )
+
+        if analysis_length_dates < 0:
+            logging.error(
+                "Proposed enrollment longer than analysis dates length:"
+                + f"{self.config.experiment.normandy_slug}"
+            )
+            raise Exception("Cannot validate experiment")
+
+        limits = TimeLimits.for_single_analysis_window(
+            last_date_full_data=self.config.experiment.end_date.strftime("%Y-%m-%d"),
+            analysis_start_days=0,
+            analysis_length_dates=analysis_length_dates,
+            first_enrollment_date=self.config.experiment.start_date.strftime("%Y-%m-%d"),
+            num_dates_enrollment=dates_enrollment,
+        )
+
+        exp = mozanalysis.experiment.Experiment(
+            experiment_slug=self.config.experiment.normandy_slug,
+            start_date=self.config.experiment.start_date.strftime("%Y-%m-%d"),
+        )
+
+        metrics = set()
+        for v in self.config.metrics.values():
+            metrics |= {m.metric for m in v}
+
+        sql = exp.build_query(
+            metrics,
+            limits,
+            "normandy",
+            self.config.experiment.enrollment_query,
+        )
+
+        dry_run_query(sql)
+
+    def run(self, current_date: datetime, dry_run: bool) -> None:
```

Comment:
Can't we just remove the `dry_run` option here? It doesn't seem to be used anywhere anymore.

Reply:
We use it in the tests here to exercise some of the mozanalysis logic: https://github.com/mozilla/jetstream/blob/main/jetstream/tests/test_analysis.py#L76

I use it in #200, too.

Calling .run() from unit tests would otherwise need a lot of mocking; we could test the private methods instead but then we're copying all of the setup that .run() does for us.

It is kinda awkward.

Reply:
I see, makes sense to keep it then